src/main/java/com/commercetools/sync/taxcategories/utils/TaxRatesUpdateActionUtils.java

```diff
@@ -112,13 +112,8 @@ private TaxRatesUpdateActionUtils() {}
                 newTaxRatesDrafts.stream()
                     .filter(
                         taxRateDraft ->
-                            Objects.equals(oldTaxRate.getCountry(), taxRateDraft.getCountry()))
-                    .filter(
-                        taxRateDraft ->
-                            oldTaxRate.getState() == null
-                                || (oldTaxRate.getState() != null
-                                    && taxRateDraft.getState() == null)
-                                || Objects.equals(oldTaxRate.getState(), taxRateDraft.getState()))
+                            Objects.equals(oldTaxRate.getCountry(), taxRateDraft.getCountry())
+                                && Objects.equals(oldTaxRate.getState(), taxRateDraft.getState()))
                     .findFirst()
                     .map(
                         matchedTaxRateDraft -> {
```

Comment:
```
newTaxRatesDrafts.remove(matchedTaxRateDraft);
```
Why do we need to remove here? Also, if we have to it seem not the best practice to modify the list we iterate over since you might have unexpected results or ConcurrentModificationException. 
https://blog.jooq.org/2014/06/13/java-8-friday-10-subtle-mistakes-when-using-the-streams-api/ (check point 8)

Reply:
@butenkor Yes, you are right we should not remove elements from collection, when we iterate through it. But in this case I don't see any problem because we are no more continuing the loop after this line(remove element from collection). We are returning back from the current loop and we will iterate with a new collection(without removed element).
> newTaxRatesDrafts.remove(matchedTaxRateDraft);
                            return singletonList(
                                ReplaceTaxRate.of(oldTaxRate.getId(), matchedTaxRateDraft));
 

And I see this is old code, we are removing elements to reduce the number of iterations every time on filter.

Reply:
Perhaps we can clone `newTaxRatesDrafts` in line 112 and iterate it, and keep the matched draft removal. It makes the code snippet look less complicated.

Reply:
> But in this case I don't see any problem because we are no more continuing the loop after this line(remove element from collection)

Since it is a bad practice you do not know which surprises it can hold for you. For example our [callbacks](https://github.com/commercetools/commercetools-sync-java/blob/1c03af337d807fce00187715705da4074cad4299/src/main/java/com/commercetools/sync/inventories/InventorySyncOptions.java#L24-L43) provide on different events a draft representation for logging/debugging purposes. Let's assume we remove all the taxRateDrafts from the original draft above, hit API error and call errorCallback - what will the user see in the taxRateDrafts array?